rasa/core/actions/action.py

```diff
@@ -101,30 +102,9 @@ def action_for_index(
             f"Domain has {domain.num_actions} actions."
         )
 
-    return action_for_name(domain.action_names[index], domain, action_endpoint)
-
-
-def action_for_name(
-    action_name: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
-) -> "Action":
-    """Create an `Action` object based on the name of the `Action`.
-
-    Args:
-        action_name: The name of the `Action`.
-        domain: The `Domain` of the current model. The domain contains the actions
-            provided by the user + the default actions.
-        action_endpoint: Can be used to run `custom_actions`
-            (e.g. using the `rasa-sdk`).
-
-    Returns:
-        The instantiated `Action` or `None` if no `Action` was found for the given
-        index.
-    """
-
-    if action_name not in domain.action_names:
-        domain.raise_action_not_found_exception(action_name)
-
-    return action_from_name(action_name, domain, action_endpoint)
+    return action_for_name_or_text(
```

Comment:
because the text of the action is the identifier for this action

Reply:
couldn't we generate an artificial name and store the text as a response on the domain?

Reply:
we could. But why? I think storing as actual text is nice and explicitly and aligns with the fact that this is exactly what TED will see

Reply:
I agree with the argument of explicitness. We have to keep both things separate in the domain anyway as we don't want to dump the end-to-end texts with the domain. And we should in my opinion also log the `ActionExecuted` event  on the tracker in a way that it's clearly reflected whether it was text or action.

Reply:
I agree with the point of explicitness - I think it conflicts with the maintainability of the bot though: 

- let's say I have e2e conversations stored in my db
- I want to change the wording for one of the e2e messages
- I'd expect this to change all occurrences of this message rather than doing it by hand in every training example
- I'd expect past conversations that I move from my conversations to my training data to use the changed message

I think this workflow is a bit tricky with the notion to store the exact text but this isn't blocking the merge.

Reply:
> I'd expect past conversations that I move from my conversations to my training data to use the changed message

you cannot do it. This is the thing. e2e is hardcoded inside Neural Network

Reply:
of course, I'd retrain after adding or changing the messages changing the model - I wouldn't expect this to work without retraining.

The thing though is that afterwards, I will still have old conversations using the old phrase which I changed. If I know to move that conversation from log -> training data it will use the old phrasing rather than the updated one.

Does that make sense?

Reply:
makes sense to me except this part. Could you maybe clarify?

> If I know to move that conversation from log -> training data it will use the old phrasing rather than the updated one.

Reply:
I'd prefer the model to use the old phrases. because imagine you changed the flow, so that you started to predict different action at this point. This is equivalent to rephrasing